### ISSUE TITLE
Incorrect prefer_const_constructors warning

### ISSUE THREAD

**Issue by spkersten** [dart-lang/linter#2150]
In the following code:

```dart
void main() {
  final foo1 = Bar({const Foo(): ""});  // 1
  final foo2 = const Bar({Foo(): ""});  // 2
}

class Foo {
  const Foo();
  
  bool operator==(dynamic other) => false;
  
  int get hashCode => 1;
}

class Bar {
  const Bar(Map<Foo, String> x);
}
```

I get a warning for the `foo1`: 

> info: Prefer const with constant constructors. (prefer_const_constructors ... 

But, with `const` at `foo2`, there is an error:

> Error compiling to JavaScript:
> main.dart:4:22:
> Error: Constant evaluation error:
>   final foo2 = const Bar({Foo(): ""});
>                      ^
> main.dart:4:27:
> Info: The key 'Foo {}' does not have a primitive operator '=='.
>  - 'Foo' is from 'main.dart'.
>   final foo2 = const Bar({Foo(): ""});
>                           ^
> Error: Compilation failed.

Dart 2.7.0

**Comment by bwilkerson** [comment]
@mraleph This is actually a bug in `analyzer`, not in the linter. Could you please move it back to the sdk repo?
